app/models/service.rb

```diff
@@ -181,29 +181,35 @@ def shutdown_guest
   end
 
   def power_states_match?(action)
-    if composite? && (power_states.uniq == map_composite_power_states(action))
-      return update_power_status(action)
-    elsif atomic? && (power_states[0] == POWER_STATE_MAP[action])
-      return update_power_status(action)
-    end
+    return update_power_status(action) if all_states_match?(action)
+    false
+  end
+
+  def all_states_match?(action)
+    return true if composite? && (power_states.uniq == map_power_states(action))
+    return true if atomic? && (power_states[0] == POWER_STATE_MAP[action])
     false
   end
 
   def composite?
-    service_template ? service_template.composite? : children.present?
+    children.present?
   end
 
   def atomic?
-    service_template ? service_template.atomic? : children.empty?
+    children.empty?
   end
 
   def orchestration_stacks
     service_resources.where(:resource_type => 'OrchestrationStack').collect(&:resource)
   end
 
-  def map_composite_power_states(action)
+  def map_power_states(action)
     action_name = "#{action}_action"
-    service_actions = service_resources.map(&action_name.to_sym).uniq
+    service_actions = [].tap do |sa|
+      each_group_resource do |svc_rsc|
+        sa << svc_rsc
+      end
+    end.map(&action_name.to_sym).uniq
```

Comment:
Maybe set `action_name` to a symbol when setting the variable `action_name = "#{action}_action".to_sym` instead of during each call.  Also, you end up calling `compact` on `service_actions` below, should that happen here?

I am feeling like this block should be broken out into a separate method.